e2e_tests/yatai_server/utils.py

```diff
@@ -0,0 +1,65 @@
+import subprocess
+import logging
+import os
+
+from bentoml import BentoService, api
+from bentoml.handlers import DataframeHandler
+from bentoml.yatai.client import YataiClient
+
+logger = logging.getLogger('bentoml.test')
+
+GRPC_PORT = '50051'
+GRPC_CHANNEL_ADDRESS = f'127.0.0.1:{GRPC_PORT}'
+
+
+class BentoServiceForYataiTest(BentoService):
+    @api(DataframeHandler)
+    def predict(self, df):
+        return 'cat'
+
+
+def delete_bento_service(bento_tag):
+    command = ['bentoml', 'delete', bento_tag, '-y']
+    proc = subprocess.Popen(
+        command,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        env=_generate_temporary_env_with_channel_address(),
+    )
+    stdout = proc.stdout.read().decode('utf-8')
+    return stdout
+
+
+def get_bento_service(bento_tag):
+    os.environ['BENTOML__YATAI_SERVICE__CHANNEL_ADDRESS'] = GRPC_CHANNEL_ADDRESS
```

Comment:
looking through this file, I think it's better to just set the env around the code to be tested, instead of doing it for every single function here.



```
import contextlib
import os


@contextlib.contextmanager
def modified_environ(*remove, **update):
    """
    Temporarily updates the ``os.environ`` dictionary in-place.

    The ``os.environ`` dictionary is updated in-place so that the modification
    is sure to work in all situations.

    :param remove: Environment variables to remove.
    :param update: Dictionary of environment variables and values to add/update.
    """
    env = os.environ
    update = update or {}
    remove = remove or []

    # List of environment variables being updated or removed.
    stomped = (set(update.keys()) | set(remove)) & set(env.keys())
    # Environment variables and values to restore on exit.
    update_after = {k: env[k] for k in stomped}
    # Environment variables and values to remove on exit.
    remove_after = frozenset(k for k in update if k not in env)

    try:
        env.update(update)
        [env.pop(k, None) for k in remove]
        yield
    finally:
        env.update(update_after)
        [env.pop(k) for k in remove_after]

@pytest.fixture()
def start_yatai_server():
     proc = None
     def func(db_url=None, store=None):
          yatai_server_command = [...]
          proc = subprocess.Popen(
             yatai_server_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE
          )
         return f'localhost:{port}'
     yeild func
     if proc:
           proc.terminate()...
  

def test_yatai_server_with_postgres_and_local_storage(get_yatai_server_url, temporary_docker_postgres_url):
    logger.info('Setting yatai server channel address to BentoML config')

    url = get_yatai_server_url()

    with modified_environ(BENTOML__YATAI_SERVICE__URL=url):
           # regular BentoService save
           # regular BentoService get and verify
           # regular BentoService run and verify
           # regular delete benoservice

```